metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,275 @@
-    print("\nCreating the pipeline definition needed to run the flow on KFP...\n")
-    print("\nCode URL of the flow to be converted to KFP: {0}\n".format(flow_code_url))
+
+    command_template_map = create_command_templates_from_graph(graph)
 
     @dsl.pipeline(
         name='MF on KFP Pipeline',
-        description='Pipeline defining KFP equivalent of the Metaflow flow'
+        description='Pipeline defining KFP equivalent of the Metaflow flow. Currently supports linear flows and flows '
+                    'with branch and join nodes'
     )
     def kfp_pipeline_from_flow():
-        """
-        This function converts the flow steps to kfp container op equivalents
-        by invoking `step_container_op` for every step in the flow and handling the order of steps.
-        """
+        S3_BUCKET = V1EnvVar(name="S3_BUCKET", value=S3_BUCKET_VALUE)
+        S3_AWS_ARN = V1EnvVar(name="S3_AWS_ARN", value=S3_AWS_ARN_VALUE)
+        pre_start_op = (pre_start_container_op())(code_url).add_env_variable(S3_BUCKET).add_env_variable(S3_AWS_ARN)
+        outputs = pre_start_op.outputs
+
+        container_op_map = {}
+        ds_root = outputs['ds_root']
+        run_id = outputs['run_id']
 
-        # Store the list of steps in reverse order
-        step_container_ops = [step_container_op(step, code_url) for step in reversed(steps)]
+        # Initial setup
+        pre_start_op.set_display_name('InitialSetup')
+        container_op_map['start'] = (step_container_op())(
+                                        command_template_map['start'],
+                                        'start',
+                                        code_url,
+                                        ds_root,
+                                        run_id
+                                        ).add_env_variable(S3_BUCKET).add_env_variable(S3_AWS_ARN)
+        container_op_map['start'].set_display_name('start')
+        container_op_map['start'].after(pre_start_op)
 
-        # Each step in the list can only be executed after the next step in the list, i.e., list[-1] is executed first, followed
-        # by list[-2] and so on.
-        for i in range(len(steps) - 1):
-            step_container_ops[i].after(step_container_ops[i + 1])
+        for step, cmd in command_template_map.items():
+            if step not in container_op_map.keys():
+                container_op_map[step] = (step_container_op())(
+                                            command_template_map[step],
+                                            step,
+                                            code_url,
+                                            ds_root,
+                                            run_id).add_env_variable(S3_BUCKET).add_env_variable(S3_AWS_ARN)
+                container_op_map[step].set_display_name(step)
+                for parent in graph.nodes[step].in_funcs:
+                    container_op_map[step].after(container_op_map[parent])
```

Comment:
Looks like you're explicitly relying on the order of the dictionary to ensure this code always works (ie. you parse parents before their children). I believe python from Python v3.6 dictionaries are insertion ordered and it's a language feature from v3.7, (see [here](https://stackoverflow.com/a/39980744/724251)) but not sure what Metaflow's supported version range is so this may not always hold true if it's outside those ranges :)

Reply:
Yes, I was relying on the insertion ordering feature. 

I just checked the metaflow channel and it looks like Taleb has had a conversation with the developers on python 2 support. Pasting below: 
> Taleb: Curious, I must not know something, why support Python2 still?
> Savin: We want to make sure we don't accidentally break users who are still on python2. Given that they can use python2 within @conda to shield themselves from any migrations, we have to make sure that metaflow core is compatible with python2, and hence the tests.

Not entirely sure what the strategy for newer code is. So just to be on the safer side, I'm decoupling the ordering from container op creation. It no longer relies on the insertion order now :)